### PR TITLE
fix(BinaryCoStream): resolve the loadAsBlob call only when the stream id ended

### PR DESCRIPTION
**Pull request by gdorsi**
Currently BinaryStream.loadAsBlob resolves into undefined if the BinaryStream isn't complete.

With this PR we change the behavoir to wait for the BinaryStream to be completed before resolving the promise.

NOTE: This is the second iteration for the bugfix. I've the first one I've also changed the behavoir of `load` but later found out that wasn't the best choice: https://github.com/gardencmp/jazz/pull/335#discussion_r1723661794

### CODE REVIEW

File: packages/cojson/src/coValues/coStream.ts [hunk 1]
```diff
@@ -96,15 +96,20 @@ export class RawCoStreamView<
...
+        }
+
+        if (streams.length > 1) {
             throw new Error(
                 "CoStream.getSingleStream() can only be called when there is exactly one stream",
             );
         }
 
-        return Object.values(this.items)[0]?.map((item) => item.value);
+        return firstStream.map((item) => item.value);
```
Comment (gdorsi) [inline packages/cojson/src/coValues/coStream.ts:112]: Done a small refactoring to avoid the ?. operator here

File: packages/cojson/src/coValues/coStream.ts [hunk 2]
```diff
@@ -306,20 +320,9 @@ export class RawBinaryCoStreamView<
             const chunk = base64URLtoBytes(
                 item.chunk.slice(binary_U_prefixLength),
             );
```
Comment (gdorsi) [inline packages/cojson/src/coValues/coStream.ts:322]: **observation:** I can see this sync decoding become a problem with large files

How do you feel about making getBinaryChunks async and see how a worklet would perform here? @aeplay

File: packages/jazz-tools/src/coValues/coStream.ts [hunk 3]
```diff
@@ -540,7 +544,28 @@ export class BinaryCoStream extends CoValueBase implements CoValue {
             allowUnfinished?: boolean;
         },
     ): Promise<Blob | undefined> {
-        const stream = await this.load(id, as, []);
+        let stream = await this.load(id, as, []);
```
Comment (gdorsi) [inline packages/jazz-tools/src/coValues/coStream.ts:547]: **question:** in which cases `stream` can be `undefined`?

Wondering what to do if the user is passing `allowUnfinished: true` and `stream` is `undefined`

File: packages/cojson/src/coValues/coStream.ts [hunk 4]
```diff
@@ -306,20 +320,9 @@ export class RawBinaryCoStreamView<
...
-            if (Date.now() - lastProgressUpdate > 100) {
-                lastProgressUpdate = Date.now();
-            }
         }
 
-        // const after = performance.now();
-        // console.log(
-        //     "getBinaryChunks bandwidth in MB/s",
-        //     (1000 * totalLength) / (after - before) / (1024 * 1024)
-        // );
```
Comment (gdorsi) [inline packages/cojson/src/coValues/coStream.ts:321]: Removed this debug code to make it easier for me to understand the logic.

Happy to revert the change if this code is still useful

File: packages/jazz-tools/src/coValues/coStream.ts (hunk 3 above)
Comment (aeplay) [reply packages/jazz-tools/src/coValues/coStream.ts:547]: When the CoValue for the stream doesn't load at all.

In that case just return undefined.

**Review by aeplay — APPROVED** [review]

**Review by aeplay — APPROVED** [review]
